### PR TITLE
Ensure Distribution can be used in Latitude and Longitude

### PR DESCRIPTION
**Pull request by mhvk**
This pull request is to address the fact that one cannot initialize a `Longitude` with a `Distribution`, which resulted from an error being raised on comparing a `Distribution` with a non-void array -- the solution was to override `__eq__` and `__ne__` (other comparisons are not affected). 

EDIT: Furthermore, for wrapping, `__getitem__` and `__setitem__` needed to be changed to ensure that things like `dist[dist < 0] += 360` work. And in `Latitude._validate_angles`, one needs to be sure that no assumption of a `float` dtype is made.

This part of a hopefully larger attempt to ensure we can use `Distribution` also inside `SkyCoord`, etc.

Note: ~since this is a bug, backporting to 5.2. But perhaps note to 5.0, since really this late it does not seem helpful to fix LTS for this any more.~ EDIT: while this is a bug, the change to `coordinates` makes me feel that it is better to treat it as an "enhancement" and not back-port.

### CODE REVIEW

**Comment by eerovaher** [comment]
It would certainly be good if `astropy.coordinates` worked better with `Distribution`, but any implementation must deal with the edge case where the `Distribution` contains values both above and below the wrapping angle. With 5631ab9ed0b6bb545b9d5c7903e363e9bdd4aa78 the following occurs:
```python
>>> from astropy import units as u
>>> from astropy.coordinates import Longitude
>>> from astropy.uncertainty import Distribution
>>> d = Distribution([359, 361] * u.deg)
>>> d.pdf_std()
<Quantity 1. deg>
>>> Longitude(d).pdf_std()
<Longitude 179. deg>
```

**Comment by mhvk** [comment]
Hmm, yes, that is unavoidable right now... Of course, the problem is that a standard deviation of a `Longitude` (and really any `Angle`) just generally doesn't make sense - one should use the circular version. The same holds for all the other `.pdf_*`.

Maybe we should have an option to use circular for them? And default to those for `Angle` subclasses? That is quite easy to implement... Let me ping @eteq as well.

p.s. Not sure we have to solve the above for this PR... @eerovaher - what do you think, shall I just open another issue?

**Comment by nstarman** [comment]
> Maybe we should have an option to use circular for them? And default to those for `Angle` subclasses? That is quite easy to implement... Let me ping @eteq as well.

That seems the right solution.
Are you thinking a kwarg to switch between linear and circular std?

**Comment by eerovaher** [comment]
> Maybe we should have an option to use circular for them? And default to those for `Angle` subclasses?

I agree that `Angle` and its subclasses should use the circular versions, but I'm not sure giving the option not to use the circular versions would be valuable.

> p.s. Not sure we have to solve the above for this PR... @eerovaher - what do you think, shall I just open another issue?

I'd rather have this sorted out in this PR, especially if the implementation is simple. The ability to create a `Longitude` from a `Distribution` is not very useful if the resulting object doesn't behave properly.

**Comment by mhvk** [comment]
I was thinking of a `kwarg` with a default that depends on either the unit or the class.

@eerovaher - I don't quite agree that `LongitudeDistribution` as is would not be useful, as it really is meant to be a step to allow `Representation` and `SkyCoord` to work - for which `pdf_mean` can be quite useful. I worry that here we are going to be stuck on implementation details and naming... (and I do think it needs a bit of thought on how exactly to implement it).

**Comment by mhvk** [comment]
Or, rephrasing, I'd rather have PRs be small and to the point. This one fixes real bugs -- of not being able to compare with `==` and `!=` and not being able to do in-place operations -- which were exposed by trying to make `LongitudeDistribution` and `LatitudeDistribution`, both of which should have worked already. 

In contract, having circular `pdf_*` is a new feature -- one that I agree is necessary, but which feels like a separate issue.

**Comment by mhvk** [comment]
Hmm, I thought we could simply use routines in `astropy.stats.circstats`, but looking closer I see that while it has `circmean`, `circstd` and `circvar`, it has no median or [s]mad or implementations for percentiles and histograms. I guess all of those involve a choice of how to sort so are less obvious.

**Comment by nstarman** [comment]
https://insidebigdata.com/2021/02/12/circular-statistics-in-python-an-intuitive-intro/
Has a suggestion for the circular median.

**Comment by mhvk** [comment]
Nice! I opened a feature request for more circular statistics, as well as discussion on how to implement it for `Distribution` -- see #14427.

File: docs/changes/uncertainty/14421.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Ensure that ``Distribution`` can be compared with ``==`` and ``!=``
+with regular arrays or scalars, and that inplace operations like
+``dist[dist<0] *= -1`` work.  This also ensures that the ``Angle`` subclasses
+``Longitude`` and ``Latitude`` can be initialized with a ``Distribution``.
```
Comment (eerovaher) [inline docs/changes/uncertainty/14421.bugfix.rst:4]: I'm not sure we want to advertise initializing `Angle` subclasses with `Distribution` quite yet.

**Review by eerovaher — COMMENTED** [review]
Let's proceed in small steps then and limit the scope of the pull request to what has already been implemented.

I do have a question about test setup. What is the advantage of implementing classes with a `setup_class()` method over using [`pytest` fixtures](https://docs.pytest.org/en/stable/how-to/fixtures.html) or perhaps global constants? There is an obvious downside to implementing test classes - they use up a level of indentation.

File: docs/changes/uncertainty/14421.bugfix.rst (hunk 1 above)
Comment (mhvk) [reply docs/changes/uncertainty/14421.bugfix.rst]: OK, you're right - let's announce that properly when the statistics make sense!

**Comment by mhvk** [comment]
On `setup_class`, it is mostly what I'm used too - I do like combining tests that (more or less) logically belong together in classes, and find the `pytest` fixtures very hard to read because the fixtures feel so unpythonic -- only with `pytest.mark.parametrize` does it make obvious sense to me what is happening...Also, with fixtures, initialization code tends to move far away from where it is used (like in `Table`), so I find I never fully understand what exactly is being tested.

Anyway, bottom line: subjective preference!

File: astropy/uncertainty/tests/test_containers.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        da2 = da + da
+        assert isinstance(da, Angle)
+        assert isinstance(da, Distribution)
+
+    @pytest.mark.parametrize("angle_cls", [Longitude, Latitude])
+    def test_pdfmean_gives_correct_subclass(self, angle_cls):
+        # Lon and Lat always fall back to Angle
+        da = angle_cls(self.dq)
+        mean = da.pdf_mean()
+        assert isinstance(mean, Angle)
```
Comment (eerovaher) [inline astropy/uncertainty/tests/test_containers.py:71]: Do we want this behavior? It is quite clear that the standard deviation of `Longitude` values should not be a `Longitude` itself, but the mean very well could be.

Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:71]: I guess this follows from the sum of longitudes being an angle. It is baked in `Longitude` right now, which I'm not touching here. 

But I agree that really this does not have to hold. Shall I just remove the isinstance check here so that we don't get unnecessary failures if we change things?

Comment (eerovaher) [reply astropy/uncertainty/tests/test_containers.py:71]: Maybe we should test `pdf_std()` instead of `pdf_mean()` because with the standard deviation we do want the class to change.

Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:71]: Good idea, adopted.

**Comment by mhvk** [comment]
@eteq - ping! This PR touches the `Distribution` class quite a bit, so review by you would be great!

**Comment by mhvk** [comment]
Since feature-freeze is coming up soon: it would be good if @eteq could review this PR, but failing that I think it should be merged; it does fix some actual bugs.

**Comment by pllim** [comment]
There is already approval and no response from @eteq in the past 2 weeks, so I am merging. Thanks!

File: astropy/coordinates/angles.py [hunk 3]
```diff
@@ -553,11 +553,17 @@ def _validate_angles(self, angles=None):
...
         elif angles.unit is u.rad:
-            limit = self.dtype.type(0.5 * np.pi)
+            limit = 0.5 * np.pi
         else:
             limit = u.degree.to(angles.unit, 90.0)
 
-        invalid_angles = np.any(angles.value < -limit) or np.any(angles.value > limit)
+        # Ensure ndim>=1 so that comparison is done using the angle dtype.
+        # Otherwise, e.g., np.array(np.pi/2, 'f4') > np.pi/2 will yield True.
+        # (This feels like a bug -- see https://github.com/numpy/numpy/issues/23247)
```
Comment (eteq) [inline astropy/coordinates/angles.py:562]: It looks like since this was written there might be a fix in a future numpy for this - might want to add a todo note aobut that

File: astropy/uncertainty/core.py [hunk 4]
```diff
@@ -311,6 +311,9 @@ def view(self, dtype=None, type=None):
 
     # Override __getitem__ so that 'samples' is returned as the sample class.
     def __getitem__(self, item):
+        if isinstance(item, Distribution):
```
Comment (eteq) [inline astropy/uncertainty/core.py:314]: I thing it might be better to use duck-typing here - will make a follow-on PR with a proposal

File: astropy/uncertainty/core.py [hunk 5]
```diff
@@ -320,6 +323,30 @@ def __getitem__(self, item):
         else:
             return result
 
+    def __setitem__(self, item, value):
+        if isinstance(item, Distribution):
```
Comment (eteq) [inline astropy/uncertainty/core.py:327]: (same here as above)

File: astropy/uncertainty/tests/test_containers.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+class TestAngles:
+    @classmethod
+    def setup_class(cls):
+        cls.a = np.arange(27.0).reshape(3, 9)
+        cls.d = Distribution(cls.a)
+        cls.q = cls.a << u.deg
+        cls.dq = Distribution(cls.q)
+
+    @pytest.mark.parametrize("angle_cls", [Angle, Longitude, Latitude])
```
Comment (eteq) [inline astropy/uncertainty/tests/test_containers.py:21]: maybe should add `EarthLocation` to this list? Although that might be a whole can of worms.

**Review by eteq — COMMENTED** [review]
Oops, should have followed-up here yesterday!  @mhvk and I discussed offline I'd prioritize this one.  But I'll just go ahead and make a PR that implements my suggestions here later today and @mhvk can review it there.

File: astropy/uncertainty/core.py (hunk 4 above)
Comment (eteq) [reply astropy/uncertainty/core.py:314]: after looking more closely I realize this is a bigger topic that's not worth worrying about right now, because it would be more consistent to do this with *all* the `isinstance` calls in this sub-package, which might have unintended consequences...

File: astropy/uncertainty/tests/test_containers.py (hunk 6 above)
Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:21]: `EarthLocation` will be interesting, since it is already a structured type. That will really need some thought, since presumably each sample then has `x,y,z` inside, but we still have to be sure `el_dist['x']` works... In other words, I'm pretty sure this will not yet work!

File: astropy/uncertainty/core.py (hunk 4 above)
Comment (mhvk) [reply astropy/uncertainty/core.py:314]: Indeed. While usually a big fan of duck-typing, I think in this instance, `isinstance` may actually be OK; e.g., I don't think we can count on other implementations to use `.distribution` as an attribute... But something to think about nevertheless...

File: astropy/coordinates/angles.py (hunk 3 above)
Comment (pllim) [reply astropy/coordinates/angles.py:562]: xref https://github.com/numpy/numpy/issues/23247

**Comment by pllim** [comment]
Ooops, my bad on merging too early. I apologize for the extra work.

**Comment by mhvk** [comment]
No worries, you couldn't know I had just decided to e-mail Erik directly...

**Comment by pllim** [comment]
Not yet... 👹
